### PR TITLE
CSHARP-3714: Implement missing test from Versioned API specification

### PR DESCRIPTION
**Pull request by MikalaiMazurenka**
JIRA: https://jira.mongodb.org/browse/CSHARP-3714
EG: https://evergreen.mongodb.com/version/60e63dec2a60ed546b45f97e

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilderExtensions.cs [hunk 1]
```diff
@@ -235,6 +235,10 @@ public static ClusterBuilder ConfigureWithConnectionString(this ClusterBuilder b
             {
                 builder = builder.ConfigureCluster(s => s.With(serverSelectionTimeout: connectionString.ServerSelectionTimeout.Value));
             }
+            if (serverApi != null)
```
Comment (MikalaiMazurenka) [inline src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilderExtensions.cs:238]: This is the only driver code change. I noticed that this method applies `serverApi` only to auth mechanisms, but should apply it to cluster builder as well.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/BulkMixedWriteOperationTests.cs [hunk 2]
```diff
@@ -1453,7 +1453,34 @@ public void WriteConcern_should_work()
         [InlineData(new[] { 10000000, 10000000, 10000000, 10000000, 7999888 }, new[] { 4, 1 })]
         public void Execute_with_multiple_deletes_should_split_batches_as_expected_when_using_write_commands_via_opmessage(int[] requestSizes, int[] expectedBatchCounts)
         {
-            RequireServer.Check().Supports(Feature.CommandMessage);
+            RequireServer.Check().Supports(Feature.CommandMessage).RequireApiVersion(false);
```
Comment (MikalaiMazurenka) [inline tests/MongoDB.Driver.Core.Tests/Core/Operations/BulkMixedWriteOperationTests.cs:1456]: The batch splitting depends on the current position in the buffer on the start of batch size calculation and `apiVersion` field wasn't initially expected here, so when present it shift the buffer position by 18, which I took into account by splitting tests with `serverApi` and without `serverApi`.

File: src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilderExtensions.cs (hunk 1 above)
Comment (jyemin) [reply src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilderExtensions.cs:238]: What is the effect of not having this?  What doesn't work?

Comment (MikalaiMazurenka) [reply src/MongoDB.Driver.Core/Core/Configuration/ClusterBuilderExtensions.cs:238]: Without this change if user want to configure `ClusterBuilder` with `serverApi` and connection string they would write:
```
var builder = new ClusterBuilder().ConfigureWithConnectionString(connectionString, serverApi);
```
But only authentication mechanism will be configured with `serverApi` and the cluster itself will have `serverApi` set to `null`. So the complete configuration will need one more call:
```
var builder = new ClusterBuilder().ConfigureWithConnectionString(connectionString, serverApi).ConfigureCluster(c => c.With(serverApi: serverApi));
```
Which I stumbled upon myself, thinking that I implemented `ConfigureWithConnectionString` to configure `serverApi` for all aspects of cluster.
So while it's debatable whether this is a bug, without this change the `ConfigureWithConnectionString` method behavior is at least unintuitive.

File: tests/MongoDB.Driver.Core.TestHelpers/CoreTestConfiguration.cs [hunk 3]
```diff
@@ -293,6 +293,31 @@ private static BuildInfoResult RunBuildInfo()
...
 
+        public static bool RequireApiVersion
+        {
+            get
+            {
+                return Environment.GetEnvironmentVariable("MONGODB_API_VERSION") != null;
+            }
+        }
+
+        public static ServerApi GetServerApi()
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Core.TestHelpers/CoreTestConfiguration.cs:304]: This should be a lazy static property like `Cluster`, `ConnectionString`, etc.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Overall LGTM. Minor changes requested.

File: tests/MongoDB.Driver.Core.Tests/Core/Operations/BulkMixedWriteOperationTests.cs (hunk 2 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Core.Tests/Core/Operations/BulkMixedWriteOperationTests.cs]: Great point!
Done

File: tests/MongoDB.Driver.Core.TestHelpers/CoreTestConfiguration.cs (hunk 3 above)
Comment (MikalaiMazurenka) [reply tests/MongoDB.Driver.Core.TestHelpers/CoreTestConfiguration.cs]: I generally think this class should be reimplemented. I do not think that overcomplicating the code with excessive usage of statics is worth the tiny performance savings. But agree that this PR is not the point to start major test code changes.
Done

**Review by MikalaiMazurenka — COMMENTED** [review]
Ready for review!

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**